### ISSUE TITLE
Cannot read MarshalAsAttribute in source generator

### ISSUE THREAD

**Issue by kant2002** [dotnet/roslyn#54754]
**Version Used**: 
3.10.0

**Steps to Reproduce**:

1. Create analyzer
2. Find desired method which has `[MarshalAs]` attribute assigned to parameter. Save to `IParameterSymbol parameterSymbol` variable.
3. Try read `parameterSymbol.GetAttributes()` 
4. This array return 0 elements, even if `[MarshalAs]` present.

Important that method should be in the separate assembly, If method is in same assembly in which source generator produce code, then `MarshalAsAttribute` appears in the list of attributes.

**Comment by PathogenDavid** [comment]
TL;DR: This is happening because `MarshalAs` isn't an attribute anymore after compilation.

I don't have a great workaround for you, but I can at least explain why it's happening (and why you might not hold your breath on this feature getting added any time soon.)

`MarshalAs` (along with `DllImport` and a few others) are not actually emitted from the compiler as attributes but as metadata directly associated with their associated method/parameter. (You can see this for yourself by looking at [the IL](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgBgAJDSA6AJQFcA7DbAWxiYCSXGFAgAHAMoiAbtjAwAzgG4qVQgGYGpJA2J0AKooxUA3lToW6AbQAiuXAJ5joGABQAiCRD5MAJvfcASgBdc0srAHlgBQhcGAwYDyiYuIS6PgUFAEMAcxgg0MpLBk1GHRgEBKgOBhQ6R2coBN9CAFZuCA5XKwBZLKgFAAss3ABBBVcAVQ4eLI5cmF99AE8xfknSELpgCFi6LMCVSgBfIA). I included an `ObsoleteAttribute` to show what a normal attribute looks like.)

Unfortunately since Roslyn doesn't have a need for marshaling info, [it doesn't parse it](https://github.com/dotnet/roslyn/blob/0049504955aeb06efe5d5960327d3663b9d1d3a2/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs#L844) when it comes from an external source.

The reason this issue doesn't come up with reflection is [the runtime synthesizes fake attributes](https://github.com/dotnet/runtime/blob/4c92aef2b08f9c4374c520e7e664a44f1ad8ce56/src/coreclr/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs#L1737-L1741). In theory Roslyn could do this too, but I'm not sure how the devs would feel about it.

----------

As for workarounds potential workarounds...

If all you need is the `UnmanagedType` of the parameter, that is stored and you should be able to get at it by reflecting [`PEParameterSymbol.MarshallingType`](https://github.com/dotnet/roslyn/blob/0049504955aeb06efe5d5960327d3663b9d1d3a2/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs#L863).

If you need anything beyond that, I believe the rest is hiding in its encoded form in [`PEParameterSymbol.MarshallingDescriptor`](https://github.com/dotnet/roslyn/blob/0049504955aeb06efe5d5960327d3663b9d1d3a2/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs#L849). I believe [`ParseNativeTypeInfo`](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/vm/mlinfo.cpp#L485) is the runtime code responsible for decoding this data so I'd use it as reference. (The format is also documented in ECMA-335 §II.23.4, but unless I'm missing something it doesn't include everything anyway. You might also find it useful to refer to [`MetaDataImport::GetMarshalAs`](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/vm/managedmdimport.cpp#L54), which is the connecting glue between `GetMarshalAsCustomAttribute` and `ParseNativeTypeInfo`.)

**Comment by kant2002** [comment]
Thanks for detailed answer. I was aware that MarshalAs is pseudo-custom attributes. And rationale why it is not included make sense. What I thinking right now, is that time changed with source generators. One use-case for source generators is implementing [efficient marshalling for .NET](https://github.com/dotnet/runtimelab/tree/feature/DllImportGenerator) and for [generation of COM RCW/CCW](https://github.com/dotnet/runtimelab/issues/306#issue-736720731) using ComWrappers. For that you may need declaration of structures and interfaces from other assemblies.

For now, I'm probably would be happy with proposed workarounds. But isn't this should be handled properly in Roslyn nowadays?

**Comment by jaredpar** [comment]
I think the best path forward here is to get an API proposal together for how we expose pseudo attributes in Roslyn. 

@333fred as I believe this has come up before. Not sure if we have an existing proposal that we can add onto here.

**Comment by 333fred** [comment]
This is similar to others, but we don't have anything specifically related to MarshalAs requested yet. We decode this information (https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/FieldSymbol.cs,291), but it is not exposed in any public form today. We'd need an API proposal for how to expose marshalling information.

**Comment by kant2002** [comment]
Initially I was thinking that somehow when calling `GetAttributes()` values from metadata would recreate `AttributeData` for appropriate IParameterSymbol/IFieldSymbol implementations. That's most intuitive way to work with these information. I do not see a value to have special API for consuming this data. If you insist that due performance implications, or maybe potential breakages it is not an option, I think API can be like that

```diff
public interface IParameterSymbol
{
+    MarshallingInformation GetMarshallingInformation();
}

public interface IFieldSymbol
{
+    MarshallingInformation GetMarshallingInformation();
}

+public struct MarshallingInformation
+{
+    public System.Runtime.InteropServices.UnmanagedType MarshalType { get; internal set; }
+    public System.Runtime.InteropServices.UnmanagedType ArrayElementType { get; internal set; }
+    public int ArraySizeConst { get; internal set; }
+    public int ArraySizeParameterIndex { get; internal set; }
+    public ISymbol CustomMarshallerType { get; internal set; }
+    public string CustomMarshallerCookie { get; internal set; }
+    public int ComInterfaceIIDParameterIndex { get; internal set; }
+}
```
Depending on the implementation of `MarshallingInformation` would be populated either from `AttributeData` or from CIL metadata.
